### PR TITLE
Add: unit tests

### PR DESCRIPTION
**Pull request by AnthonyMadia**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [x] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

- installed `msw` for mocking [API](https://mswjs.io/#:~:text=API%20mocking%20of%20the%20next%20generation) calls
- [installed](https://github.com/github/fetch#:~:text=The%20fetch()%20function,traditional%20web%20applications) `whatwg-fetch` to mock fetch and eliminate `NODE` option line in `package.json` to stop any reference fetch errors that node was throwing on tests. This is a more elegant solution than having the option flag in config. 
- create abstractions for render functions to reduce code in tests — I am  continuing to scan for ways to reduce repeat code.
- move `__utils__` folder to `__tests__` dir
- **Unit Tests** for:
    - postBlock
    - commentList
    - postsList
    - commentField
    - writingField
    - voteBox
    - loginPage
    - mainPage
    - basicPage
    - newPage
    - postPage
- **Note**: will add improved assertions for network requests using the MSW library.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Some tests are currently being skipped because I am researching a better implementation to mock API calls.

### CODE REVIEW

File: packages/frontend/src/__tests__/mainPage.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    renderMainPage(userData, unirepData, postData)
+    expect(
+        screen.getByText(/community built on ideas, not identities./i)
+    ).toBeInTheDocument()
+    expect(
+        screen.getByText(/stay up to date & share everything with everyone./i)
+    ).toBeInTheDocument()
+    // userState is *true*
+    expect(screen.getByText(/my rep/i)).toBeInTheDocument()
+    expect(screen.getByText(/30/i)).toBeInTheDocument()
```
Comment (vimwitch) [inline packages/frontend/src/__tests__/mainPage.js:100]: Would be nice to use `unirepData.postReputation` here (I think that's what this 30 is). We should at least have a comment saying what this number is.

File: packages/frontend/src/__tests__/mainPage.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                content: 'mocked post content',
+            },
+        },
+        loadFeed: jest.fn(),
+    }
+    renderMainPage(userData, unirepData, postData)
+    // simulate user clicking on the 'create post' link
+    const createPostLink = screen.getByText(/create post/i)
+    await userEvent.click(createPostLink)
+    expect(screen.getByText(/new/i)).toBeInTheDocument()
```
Comment (vimwitch) [inline packages/frontend/src/__tests__/mainPage.js:134]: This assertion is so generic 😅

**Review by vimwitch — CHANGES_REQUESTED** [review]
Looks like there are some good fixes in here. In half the test files you [create some shared variables at the top level scope](https://github.com/Unirep/Unirep-Social/pull/26/files#diff-3bf576dc272d073919effc20e6951b84903bec10f58ea4bcda8dd600ef0ecc1eR5) and in the other half of the tests you [make local variables in each test](https://github.com/Unirep/Unirep-Social/pull/26/files#diff-8c1e8a072e4502a4863170cf2683d895b99d02c2649d74e8afc1036feabec819R18). Definitely go with the second approach. Shared state between tests is very difficult to maintain and debug.

Also just wondering, why are some of the tests skipped? Are they just WIP and you're periodically requesting review/merge as you go?

**Comment by vimwitch** [comment]
Also @kittybest we should look for a solution for [this problem](https://app.circleci.com/pipelines/github/Unirep/Unirep-Social/81/workflows/93cb636c-3937-4a71-a879-a73cdf9e5efd/jobs/402?invite=true#step-106-75).

File: packages/frontend/src/__tests__/commentBlock.js [hunk 3]
```diff
@@ -1,27 +1,93 @@
...
+            },
+        },
+    }
+
+    const userData = {
+        userState: 'userState',
+        currentEpochKeys: ['user epoch_key test'],
+    }
+
+    renderCommentField(userData, postData, page, post, mockedCloseComment)
```
[the remark this answers is not in the document]
Comment (AnthonyMadia) [reply packages/frontend/src/__tests__/commentBlock.js:65]: Yep, will do -- some of this code needed to be refactored.

**Comment by AnthonyMadia** [comment]
Some of the tests are [skipped](https://github.com/Unirep/Unirep-Social/pull/26#:~:text=Some%20tests%20are%20currently%20being%20skipped%20because%20I%20am%20researching%20a%20better%20implementation%20to%20mock%20API%20calls.) because of the networking errors I was having when trying to mock the API calls. I would get `Error: Error: connect ECONNREFUSED 127.0.0.1:3001` errors. The tests would still pass but the error logs to console. So some tests are a WIP and thought it would be good to get feedback before moving on. 

Also, I would not necessarily view these unit tests as final because the assertions are still being improved. They build a good foundation to build upon as time goes on. Pushing changes now.

**Comment by AnthonyMadia** [comment]
Also, if you want to see the errors I was getting, just take the .`skip `out of the last test on `mainPage.js` and run the test. I tried mocking a server with MSW but that was not working so may be an issue with how I am handling asynchronous code. @vimwitch

File: packages/frontend/src/__tests__/mainPage.js (hunk 2 above)
Comment (AnthonyMadia) [reply packages/frontend/src/__tests__/mainPage.js:132]: This test is WIP.

**Comment by kittybest** [comment]
> Also, if you want to see the errors I was getting, just take the .`skip `out of the last test on `mainPage.js` and run the test. I tried mocking a server with MSW but that was not working so may be an issue with how I am handling asynchronous code. @vimwitch

What's the problem if mocking a server with MSW?

**Comment by AnthonyMadia** [comment]
> 

Try removing the `.skip` running the last test on `mainPage.js` and you should see a connection error log to the console. I am not sure where that issue is stemming from. Initially, I thought I had to mock a server, but that does not seem to be the case. @kittybest

**Comment by AnthonyMadia** [comment]
Disregard the milestones.

**Comment by vimwitch** [comment]
The problem is twofold. First, msw seems to [need an absolute url to match in nodejs](https://mswjs.io/docs/getting-started/integrate/node#direct-usage). Second, the url in the config wasn't being properly set in the tests.

I added a check for `NODE_ENV` that sets the value for `SERVER` and `DEFAULT_ETH_PROVIDER_URL`. I'm not sure how much of these we should mock, probably enough to get basic tests working. Also not sure how much of the contexts we want/need to test.

**Comment by AnthonyMadia** [comment]
> The problem is twofold. First, msw seems to [need an absolute url to match in nodejs](https://mswjs.io/docs/getting-started/integrate/node#direct-usage). Second, the url in the config wasn't being properly set in the tests.

Ah, okay. This is interesting because I was using absolute paths before and then tried without to make sure. Does this make all the tests pass? I still see pending for the CI frontend-test. 

And I do not think testing context would be the best use of time. That gets down to testing implementation details and frontend tests should really focus on how the user interacts with the app.

**Comment by vimwitch** [comment]
I forgot one other thing. The server needs to be mocked before the files are imported. The synchronizer is started on import so if you mock after import you'll miss some requests. Not great.

> And I do not think testing context would be the best use of time. That gets down to testing implementation details and frontend tests should really focus on how the user interacts with the app.

Yea I agree

> I still see pending for the CI frontend-test.

I was debugging a failure in the ci with ssh :P

**Comment by vimwitch** [comment]
Ah it's actually still failing. Not sure why

**Comment by AnthonyMadia** [comment]
> Ah it's actually still failing. Not sure why

Yea, looks like when a test is testing asynchronous behavior (currently using async await), the tests fail. This is happening on `mainPage.js` and `loginPage.js`. Both throw the error: `Error: Error: getaddrinfo ENOTFOUND testurl.invalidtld`

**Comment by vimwitch** [comment]
@AnthonyMadia Does it pass for you locally? The tests are fine on my machine

**Comment by AnthonyMadia** [comment]
@vimwitch Nope, I get the errors I posted after running tests on this PR. And how are your tests passing locally and the CI has them failing?

**Comment by vimwitch** [comment]
Absolutely no idea

**Comment by AnthonyMadia** [comment]
If they pass, I'll get this PR locally and try again

**Comment by vimwitch** [comment]
Ok finally cracked it. Just needed to start msw before the `beforeAll` hook, e.g. on import.

**Comment by AnthonyMadia** [comment]
Nice! Just tried a few async userEvents and they work! This is great.

**Pull request merged**